File: wicket-devutils/src/main/java/org/apache/wicket/devutils/debugbar/WicketDebugBar.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.wicket.devutils.debugbar;

import java.util.ArrayList;
import java.util.List;

import org.apache.wicket.Application;
import org.apache.wicket.MetaDataKey;
import org.apache.wicket.ResourceReference;
import org.apache.wicket.devutils.DevUtilsPanel;
import org.apache.wicket.markup.html.CSSPackageResource;
import org.apache.wicket.markup.html.JavascriptPackageResource;
import org.apache.wicket.markup.html.image.Image;
import org.apache.wicket.markup.html.list.ListItem;
import org.apache.wicket.markup.html.list.ListView;

/**
 * 
 * @author Jeremy Thomerson <jthomerson@apache.org>
 */
public class WicketDebugBar extends DevUtilsPanel {

	private static final MetaDataKey<List<IDebugBarContributor>> CONTRIBS_META_KEY = new MetaDataKey<List<IDebugBarContributor>>() {
		private static final long serialVersionUID = 1L;
	};

	static {
		registerStandardContributors();
	}

	private static final long serialVersionUID = 1L;

	public WicketDebugBar(String id) {
		super(id);
		add(CSSPackageResource.getHeaderContribution(WicketDebugBar.class,
				"wicket-debugbar.css"));
		add(JavascriptPackageResource.getHeaderContribution(
				WicketDebugBar.class, "wicket-debugbar.js"));
		add(new Image("logo", new ResourceReference(WicketDebugBar.class,
				"wicket.png")));
		add(new Image("removeImg", new ResourceReference(WicketDebugBar.class,
				"remove.png")));
		List<IDebugBarContributor> contributors = getContributors();
		if (contributors.isEmpty()) {
			// we do this so that if you have multiple applications running in the same container,
			//	each ends up registering its' own contributors (wicket-examples for example)
			registerStandardContributors();
			contributors = getContributors();
		}
		add(new ListView<IDebugBarContributor>("contributors", contributors) {
			private static final long serialVersionUID = 1L;

			@Override
			protected void populateItem(ListItem<IDebugBarContributor> item) {
				item.add(item.getModelObject().createComponent("contrib",
						WicketDebugBar.this));
			}
		});
	}
	
	@Override
	public boolean isVisible() {
		return getApplication().getDebugSettings().isDevelopmentUtilitiesEnabled();
	}

	/**
	 * Register your own custom contributor that will be part of the debug bar.
	 * You must have the context of an application for this thread at the time
	 * of calling this method.
	 * 
	 * @param contrib
	 *            custom contributor - can not be null
	 */
	public static void registerContributor(IDebugBarContributor contrib) {
		if (contrib == null) {
			throw new IllegalArgumentException("contrib can not be null");
		}

		List<IDebugBarContributor> contributors = getContributors();
		contributors.add(contrib);
		Application.get().setMetaData(CONTRIBS_META_KEY, contributors);
	}

	private static List<IDebugBarContributor> getContributors() {
		List<IDebugBarContributor> list = Application.get().getMetaData(
				CONTRIBS_META_KEY);
		return list == null ? new ArrayList<IDebugBarContributor>() : list;
	}

	private static void registerStandardContributors() {
		registerContributor(VersionDebugContributor.DEBUG_BAR_CONTRIB);
		registerContributor(InspectorDebugPanel.DEBUG_BAR_CONTRIB);		
		registerContributor(SessionSizeDebugPanel.DEBUG_BAR_CONTRIB);		
	}
}

File: wicket-devutils/src/main/java/org/apache/wicket/devutils/debugbar/wicket-debugbar.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
;(function (undefined) {

	'use strict';

	if (typeof(Wicket) === 'undefined') {
		window.Wicket = {};
	}

	if (typeof(Wicket.debugBar) === 'object') {
		return;
	}

	function isVisible(element) {
		return !!(element.offsetWidth || element.offsetHeight || element.getClientRects().length);
	}

	Wicket.debugBar = function() {

		function setExpandedCookie(value) {
			document.cookie =  "wicketDebugBarState=" + window.encodeURIComponent(value);
		}

		function getExpandedCookie() {
			const name = 'wicketDebugBarState';
			if (document.cookie.length > 0) {
				let start = document.cookie.indexOf (name + "=");
				if (start !== -1) {
					start = start + name.length + 1;
					let end = document.cookie.indexOf(";", start);
					if (end === -1) {
						end = document.cookie.length;
					}
					return window.decodeURIComponent(document.cookie.substring(start,end));
				} else {
					return null;
				}
			} else {
				return null;
			}
		}

		const collapse = document.getElementById('wicketDebugBarCollapse');
		if (collapse) {
			collapse.addEventListener("click", function() {
				const content = document.getElementById('wicketDebugBarContents');
				const wasVisible = isVisible(content);
				setExpandedCookie(!wasVisible);
				if (wasVisible) {
					Wicket.DOM.hide(content);
				} else {
					Wicket.DOM.show(content);
				}
			});
		}

		const remove = document.getElementById('wicketDebugBarRemove');
		if (remove) {
			remove.addEventListener("click", function() {
				const bar = document.getElementById('wicketDebugBar');
				setExpandedCookie(!isVisible(bar));
				Wicket.DOM.hide(bar);
			});
		}

	    // determine state and set it
		if (getExpandedCookie() === 'false') {
			Wicket.DOM.hide(document.getElementById('wicketDebugBarContents'));
		}

		const original = Wicket.Log.error;
		Wicket.Log.error = function() {
			original.apply(Wicket.Log, arguments);

			const bar = document.getElementById('wicketDebugBar');
			if (bar) {
				bar.classList.add('wicketDebugBarError');
				bar.addEventListener('animationend', function onAnimationEnd() {
					bar.classList.remove('wicketDebugBarError');
					bar.removeEventListener('animationend', onAnimationEnd);
				}, { once: true });
			}
		};
	};
})();
